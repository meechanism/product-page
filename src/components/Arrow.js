import React from 'react';
import styled from 'styled-components';

import Colors from '../styles/colors';
import MediaQuery from '../styles/media-queries';

const sharedStyles = `
  position: absolute;
  top: 50%;
  width: 3vmin;
  height: 3vmin;
  background: transparent;
  border-top: 2px solid black;
  border-right: 2px solid black;
  box-shadow: 0 0 0 lightgray;
  transition: all 200ms ease;

  &:before {
      content: '';
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-40%,-60%) rotate(45deg);
      width: 200%;
      height: 200%;
  }
`;

export const IconLeft = styled.i`
  ${sharedStyles}
  left: 0.75rem;
  transform: translate3d(0, -50%, 0) rotate(-135deg);
`;

export const IconRight = styled.i`
  ${sharedStyles}
  right: 0.75rem;
  transform: translate3d(0, -50%, 0) rotate(45deg);
`;

export const ArrowContainer = styled.a`
  opacity: 0.75;
  font-size: 2rem;
  position: absolute;
  top: 50%;
  padding: 1rem 0.5rem;
  ${(props) => (props.right ? 'right: 1rem' : null)};
  margin: 0;
  width: 3vmin;
  height: 3vmin;
  background: white;
  color: black;
  &:hover,
  &:focus {
    opacity: 1;
    color: ${Colors.primary};
    i {
      border-color: ${Colors.primary};
    }
  }

  ${MediaQuery.small`
    ${(props) => (props.right ? 'right: 0' : null)};
  `}
`;

export const ArrowLeft = ({ onClick }) => (
  <ArrowContainer href="" onClick={onClick}>
    <IconLeft />
  </ArrowContainer>
);

export const ArrowRight = ({ onClick }) => (
  <ArrowContainer href="" right onClick={onClick}>
    <IconRight />
  </ArrowContainer>
);
