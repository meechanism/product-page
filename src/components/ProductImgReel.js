import React from 'react';
import styled from 'styled-components';

import Colors from '../styles/colors';

const ImgReelWrapper = styled.div`
  margin: 0;
  display: flex;
  justify-content: center;
`;

const Img = styled.img`
  width: 100%;
  max-height: 80px;
  border-radius: 2px;
`;

export const ImgWrapper = styled.a`
  display: block;
  width: 95px;
  padding: 0.5rem;
  &:hover {
    cursor: pointer;
    opacity: 0.75;
  }
  border-radius: 2px;
  border: 1px solid ${(props) => (props.active ? Colors.gray : 'transparent')};
`;

export const saveImg = (save, index) => (e) => {
  e && e.preventDefault();
  save(index);
};

const ProductImgReel = ({ images, setCurrImgIndex, currImgIndex }) => (
  <ImgReelWrapper>
    {images.map((currImg, i) => {
      const key = `product-image-${i}`;
      return (
        <ImgWrapper
          href=""
          key={key}
          onClick={saveImg(setCurrImgIndex, i)}
          active={currImgIndex === i}>
          <Img src={currImg.href} alt={key} />
        </ImgWrapper>
      );
    })}
  </ImgReelWrapper>
);

export default ProductImgReel;
