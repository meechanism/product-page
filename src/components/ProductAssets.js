import React, { useState } from 'react';
import styled from 'styled-components';

import MediaQuery from '../styles/media-queries';
import ProductImgReel from './ProductImgReel';
import { ArrowLeft, ArrowRight } from './Arrow';

const ProductAssetsWrapper = styled.div`
  margin: ${(props) => (props.compact ? '1rem' : '2rem 4rem 2rem 0')};
  ${MediaQuery.medium`
    margin: 1rem;
  `};

  ${MediaQuery.small`
    margin: 0 0 1rem;
  `};
`;

const HeroImg = styled.img`
  width: 100%;
  border-radius: 2px;
  margin: ${(props) => (props.compact ? '0 0 1rem ' : '0 0 2rem')};
`;

const getNext = (save, index, totalAssets) => (e) => {
  e.preventDefault();
  if (index === totalAssets - 1) {
    save(0);
  } else {
    save(index + 1);
  }
};

const getPrev = (save, index, totalAssets) => (e) => {
  e.preventDefault();
  if (index === 0) {
    save(totalAssets - 1);
  } else {
    save(index - 1);
  }
};

const ProductAssets = ({ hero, images, name, compact }) => {
  const [currImgIndex, setCurrImgIndex] = useState(0);
  const allAssets = [hero, ...images];
  const totalAssets = allAssets.length;
  return (
    <ProductAssetsWrapper compact={compact}>
      {compact && (
        <>
          <ArrowLeft
            href=""
            onClick={getPrev(setCurrImgIndex, currImgIndex, totalAssets)}
          />
          <ArrowRight
            href=""
            onClick={getNext(setCurrImgIndex, currImgIndex, totalAssets)}
          />
        </>
      )}
      <HeroImg
        compact={compact}
        src={allAssets[currImgIndex].href}
        alt={name}
      />
      {totalAssets > 1 ? (
        <ProductImgReel
          compact={compact}
          name={name}
          images={allAssets}
          setCurrImgIndex={setCurrImgIndex}
          currImgIndex={currImgIndex}
        />
      ) : null}
    </ProductAssetsWrapper>
  );
};

export default ProductAssets;
