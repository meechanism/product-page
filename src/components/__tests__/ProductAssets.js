import React from 'react';
import ProductAssets, { HeroImg, getNext, getPrev } from '../ProductAssets';
import productData from '../../data/products';
import ProductImgReel from '../ProductImgReel';
import { ArrowLeft, ArrowRight } from '../Arrow';

describe('ProductAssets component', () => {
  const testProduct = productData.groups[0];
  const { hero, images, name, compact } = testProduct;
  const props = { hero, images, name, compact };
  const saveFn = jest.fn();

  const e = { preventDefault: jest.fn() };

  it('should render', () => {
    const output = shallow(<ProductAssets {...props} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render hero img', () => {
    const output = shallow(<ProductAssets {...props} compact={true} />);
    const img = output.find(HeroImg);
    expect(img.exists()).toBeTruthy();
    expect(img.props().src).toBe(hero.href);
    expect(img.props().alt).toBe(name);
  });

  it('should render product image reel', () => {
    const output = shallow(<ProductAssets {...props} compact={true} />);
    const imgReel = output.find(ProductImgReel);
    expect(imgReel.exists()).toBeTruthy();
  });

  it('should render not render product image reel when there is less than 2 photos', () => {
    const newProps = { hero, name, compact, images: [] };
    const output = shallow(<ProductAssets {...newProps} compact={true} />);
    const imgReel = output.find(ProductImgReel);
    expect(imgReel.exists()).toBeFalsy();
  });

  it('should render arrows for compact variation', () => {
    const output = shallow(<ProductAssets {...props} compact={true} />);
    expect(output.find(ArrowLeft).exists()).toBeTruthy();
    expect(output.find(ArrowRight).exists()).toBeTruthy();
  });

  describe('getNext', () => {
    it('should increment to next index', () => {
      getNext(saveFn, 1, 5)(e);
      expect(e.preventDefault).toHaveBeenCalled();
      expect(saveFn).toHaveBeenCalledWith(2);
    });

    it('should increment back to first item if last index clicked', () => {
      getNext(saveFn, 4, 5)(e);
      expect(e.preventDefault).toHaveBeenCalled();
      expect(saveFn).toHaveBeenCalledWith(0);
    });
  });

  describe('getPrev', () => {
    it('should increment to prev index', () => {
      getPrev(saveFn, 1, 5)(e);
      expect(e.preventDefault).toHaveBeenCalled();
      expect(saveFn).toHaveBeenCalledWith(0);
    });

    it('should increment to last item if first index clicked', () => {
      getPrev(saveFn, 0, 5)(e);
      expect(e.preventDefault).toHaveBeenCalled();
      expect(saveFn).toHaveBeenCalledWith(4);
    });
  });
});
