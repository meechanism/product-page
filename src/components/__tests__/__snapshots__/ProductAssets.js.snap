// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductAssets component should render 1`] = `
<styled.div>
  <styled.img
    alt="Belgian Linen Ladder Stripe Embroidery Duvet Cover &amp; Shams - Slate/White"
    src="https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg"
  />
  <ProductImgReel
    currImgIndex={0}
    images={
      Array [
        Object {
          "alt": "",
          "height": 363,
          "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg",
          "meta": "",
          "rel": "hero",
          "size": "m",
          "width": 363,
        },
        Object {
          "alt": "",
          "height": 363,
          "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg",
          "meta": "",
          "rel": "althero",
          "size": "m",
          "width": 363,
        },
        Object {
          "alt": "",
          "height": 363,
          "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0012/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-m.jpg",
          "meta": "",
          "rel": "althero",
          "size": "m",
          "width": 363,
        },
      ]
    }
    name="Belgian Linen Ladder Stripe Embroidery Duvet Cover &amp; Shams - Slate/White"
    setCurrImgIndex={[Function]}
  />
</styled.div>
`;
